channeldb/control_tower.go

```diff
@@ -0,0 +1,410 @@
+		// this may be a duplicate call.
+		case paymentStatus == StatusCompleted:
+			updateErr = ErrPaymentAlreadyCompleted
+			return nil
+
+		default:
+			updateErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// Record the successful payment info atomically to the
+		// payments record.
+		err = bucket.Put(paymentSettleInfoKey, preimage[:])
+		if err != nil {
+			return err
+		}
+
+		return bucket.Put(paymentStatusKey, StatusCompleted.Bytes())
+	})
+	if err != nil {
+		return err
+	}
+
+	return updateErr
+
+}
+
+// Fail transitions a payment into the Failed state. After invoking this
+// method, InitPayment should return nil on its next call for this payment
+// hash, allowing the switch to make a subsequent payment.
+func (p *paymentControl) Fail(paymentHash lntypes.Hash) error {
+	var updateErr error
+	err := p.db.Batch(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		// Reset the update error, to avoid carrying over an error
+		// from a previous execution of the batched db transaction.
+		updateErr = nil
+
+		switch {
+
+		// Our records show the payment as still being grounded,
+		// meaning it never should have left the switch.
+		case paymentStatus == StatusGrounded:
+			updateErr = ErrPaymentNotInitiated
+			return nil
+
+		// A failed response was received for an InFlight payment, mark
+		// it as Failed to allow subsequent attempts.
+		case paymentStatus == StatusInFlight:
+
+		// The payment was completed previously, and we are now
+		// reporting that it has failed. Leave the status as completed,
+		// but alert the user that something is wrong.
+		case paymentStatus == StatusCompleted:
+			updateErr = ErrPaymentAlreadyCompleted
+			return nil
+
+		// The payment was already failed, and we are now reporting that
+		// it has failed again. Leave the status as failed, but alert
+		// the user that something is wrong.
+		case paymentStatus == StatusFailed:
+			updateErr = ErrPaymentAlreadyFailed
+			return nil
+
+		default:
+			updateErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// A failed response was received for an InFlight payment, mark
+		// it as Failed to allow subsequent attempts.
+		return bucket.Put(paymentStatusKey, StatusFailed.Bytes())
+	})
+	if err != nil {
+		return err
+	}
+
+	return updateErr
+}
+
+// fetchPaymentBucket fetches or creates the sub-bucket assigned to this
+// payment hash.
+func fetchPaymentBucket(tx *bbolt.Tx, paymentHash lntypes.Hash) (
+	*bbolt.Bucket, error) {
+
+	payments, err := tx.CreateBucketIfNotExists(sentPaymentsBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	return payments.CreateBucketIfNotExists(paymentHash[:])
+}
+
+// fetchPaymentStatus fetches the payment status from the bucket.  If the
+// status isn't found, it will default to "StatusGrounded".
+func fetchPaymentStatus(bucket *bbolt.Bucket) PaymentStatus {
```

Comment:
Is storing paymentStatus in the database actually necessary? If there the settleinfo key is present, the outcome is success, if the failinfo key is there, the outcome is fail and otherwise the payment is in flight.

Reply:
I am also thinking now, why do we use those bolt keys anyway? If the complete payment state including attempt is serialized as a single value of the payment hash key, there is less that can go wrong. (like having both settle and fail info at some point).

Reply:
> Is storing paymentStatus in the database actually necessary? If the the settleinfo key is present, the outcome is success, if the failinfo key is there, the outcome is fail and otherwise the payment is in flight.

With the addition of `failInfo`, you are right, the status can be inferred from the available settle/fail data. So potentially we don't have to store them explicitly in the DB. However, the status codes are nice to expose to the caller/RPC. I see no problem deferring any removal of the statuses from the DB for later though.

> I am also thinking now, why do we use those bolt keys anyway? If the complete payment state including attempt is serialized as a single value of the payment hash key, there is less that can go wrong. (like having both settle and fail info at some point).

This was a choice I made to avoid having to deserialize/serialize the whole struct every time we were only modifying certain fields of the payment. Felt more natural to be able to only alter the buckets with the data being changed. Seraialization is also less complex when not having to deal with optional fields.

Reply:
>I see no problem deferring any removal of the statuses from the DB for later though.

Why would you want to do that? Introduce it here even though it is not needed and remove it again later?

Reply:
It's just a matter of it being explicit vs implicit, myself I favor being explicit whenever possible, as it's much clearer from the PoV of code IMO.

Reply:
The diff is not big, I'm okay with just inferring it: https://github.com/lightningnetwork/lnd/pull/2761/commits/1495d296a3a47a57ea5746729832e311fee981e3